### ISSUE TITLE
How to adjoint the barotropic gyre model from Oceananigans

### ISSUE THREAD

**Issue by swilliamson7** [DJ4Earth/DJ-barotropic-gyre#1]
As of right now, the steps to get to an adjoint-able barotropic gyre model are

1) Create a "forward step" function. Right now the barotropic gyre model in Oceananigans takes in some initial inputs, and then just runs the model forward in its entirety to return the fields $u,$ $v,$ and $\eta$ at the end, if understood correctly. Ideally, we would instead take just one forward step:
              $$f(\mathbf{u}(t), \mathbf{v}(t), \mathbf{\eta}(t)) = (\mathbf{u}(t + \Delta t), \mathbf{v}(t + \Delta t), \mathbf{\eta}(t + \Delta t) )$$
(not trying to say $f$ returns a vector, just returns the fields at the next time step.) 

2) Once we have that forward function, we first want to apply Enzyme to see if it works. This shouldn't be difficult at this point (hopefully). This step will also let us see if there are issues with Enzyme + Oceananigans that would need to be worked out. 

3) Then we can run a trial sensitivity analysis. Namely, per Patrick's suggestion we can choose the cost function $$J(\mathbf{u}(t_f), \mathbf{v}(t_f)) = \mathbf{u}^2(t_f) + \mathbf{v}^2(t_f) = \sum_{j, k} u_{jk}^2 + v_{jk}^2$$ (i.e. we sum $u^2 + v^2$ at every point on the grid and add them all together.) Then we can ask what happens to this quantity if we change the initial $u$ field a little bit. This will just be running the "adjoint equations" step-by-step backwards.

**Comment by swilliamson7** [comment]
@vchuravy maybe you would know how we can modify the Oceananigans code to complete my steps? There might be a better way to use the code than to explicitly create some forward step like I described in (1), I mainly just want something I can actually give to Enzyme

**Comment by NoraLoose** [comment]
@glwagner may also be able to help us here!

**Comment by glwagner** [comment]
I'm happy to help! You may not need to clone Oceananigans (unless you have to modify the source code). I can submit a PR if you like! There is a function that takes one time-step:

```julia
using Oceananigans: time_step!

# create model

time_step!(model, dt)
```

which I can include in the PR.

**Comment by swilliamson7** [comment]
@glwagner that would be helpful, thank you! Do I need to add you to the repo to allow you to create a PR?

**Comment by glwagner** [comment]
You don't! I can submit one from a fork.

**Comment by vchuravy** [comment]
So the first step would be to just focus on the CPU version of Oceanigans :) @wsmoses is slowly working towards custom rules so we might be able to support FFTs soon enough.

**Comment by swilliamson7** [comment]
@vchuravy What's an FFT? Also, I (briefly) tried taking Greg's function that does one step (thanks a bunch to Greg!) and differentiating it with Enzyme yesterday. I ran into a question, the velocities are stored in an offset array, is Enzyme compatible with this? Will I need to provide an "offset array" placeholder for Enzyme to put the gradient in? I might open a new issue here to discuss this, but I think that the function I have should be all I need to test Enzyme.

**Comment by glwagner** [comment]
FFT: Fast Fourier Transform. I can update the model configuration so that it doesn't use an FFT (I think one of the changes I made reverted to an FFT-based solver for the free surface). We can also make the model fully explicit if that would help.

**Comment by vchuravy** [comment]
Yeah OffsetArray should be fine, but you need to provide the same structural type as the adjoint.
